tests/src/junit/java/net/neoforged/neoforge/unittest/transfer/ItemTests.java

```diff
@@ -0,0 +1,261 @@
+
+        checkComparatorOutput(testContainer);
+
+        // Check that we return sensible results if amount stored > capacity
+        ItemStack oversizedStack = new ItemStack(Items.DIAMOND_PICKAXE, 2);
+        SimpleContainer simpleContainer = new SimpleContainer(oversizedStack);
+        var wrapper = VanillaContainerWrapper.of(simpleContainer);
+
+        try (Transaction transaction = Transaction.open(null)) {
+            assertEquals(0L, wrapper.insert(ItemResource.of(oversizedStack), 10, transaction));
+            transaction.commit();
+        }
+    }
+
+    private static boolean stackEquals(ItemStack stack, Item item, int count) {
+        return stackEquals(stack, ItemResource.of(item), count);
+    }
+
+    private static boolean stackEquals(ItemStack stack, ItemResource variant, int count) {
+        return variant.matches(stack) && stack.getCount() == count;
+    }
+
+    private static class TestWorldlyContainer extends SimpleContainer implements WorldlyContainer {
+        private static final int[] SLOTS = new int[] { 0, 1, 2 };
+
+        TestWorldlyContainer() {
+            super(SLOTS.length);
+        }
+
+        @Override
+        public int[] getSlotsForFace(Direction face) {
+            return SLOTS;
+        }
+
+        @Override
+        public boolean canPlaceItem(int slot, ItemStack stack) {
+            return slot != 0 || !stack.is(Items.BUCKET); // can't have buckets in slot 0.
+        }
+
+        @Override
+        public boolean canPlaceItemThroughFace(int slot, ItemStack stack, @Nullable Direction dir) {
+            return dir != Direction.DOWN;
+        }
+
+        @Override
+        public boolean canTakeItemThroughFace(int slot, ItemStack stack, Direction dir) {
+            return dir != Direction.UP;
+        }
+    }
+
+    /**
+     * Test insertion when {@link Container#getMaxStackSize()} is the bottleneck.
+     */
+    @Test
+    public void testLimitedStackCountContainer() {
+        ItemResource diamond = ItemResource.of(Items.DIAMOND);
+        LimitedStackCountContainer container = new LimitedStackCountContainer(diamond.toStack(), diamond.toStack(), diamond.toStack());
+        var wrapper = VanillaContainerWrapper.of(container);
+
+        // Should only be able to insert 2 diamonds per stack * 3 stacks = 6 diamonds.
+        try (Transaction transaction = Transaction.open(null)) {
+            if (wrapper.insert(diamond, 1000, transaction) != 6) {
+                throw new AssertionError("Only 6 diamonds should have been inserted.");
+            }
+
+            checkComparatorOutput(container);
+        }
+    }
+
+    /**
+     * Test insertion when {@link Item#getMaxStackSize} is the bottleneck.
+     */
+    @Test
+    public void testLimitedStackCountItem() {
+        ItemResource diamondPickaxe = ItemResource.of(Items.DIAMOND_PICKAXE);
+        LimitedStackCountContainer container = new LimitedStackCountContainer(5);
```

Comment:
Is there a reason this doesn't just use `SimpleContainer` given it is explicitly checking the `Item#getMaxStackSize` is the bottleneck, and not caring about the restriction put on `Container#getMaxStackSize`